fx quant: hook up ConvTranspose{n}d

vkuzo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #49740 fx quant: split linear test cases
* **#49717 fx quant: hook up ConvTranspose{n}d**

Summary:

Quantization of `ConvTranpose{n}d` is supported in Eager mode. This PR
adds the support for FX graph mode.

Note: this currenlty only works in `qnnpack` because per-channel weights
are not supported by quantized conv transpose. In a future PR we should throw
an error when someone tries to quantize a ConvTranspose model with per-channel
weight observers until this is fixed.

Test Plan:

```
python test/test_quantization.py TestQuantizeFxOps.test_conv_transpose_1d
python test/test_quantization.py TestQuantizeFxOps.test_conv_transpose_2d
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D25674636](https://our.internmc.facebook.com/intern/diff/D25674636)

vkuzo committed:
fx quant: hook up ConvTranspose{n}d

Summary:

Quantization of `ConvTranpose{n}d` is supported in Eager mode. This PR
adds the support for FX graph mode.

Note: this currenlty only works in `qnnpack` because per-channel weights
are not supported by quantized conv transpose. In a future PR we should throw
an error when someone tries to quantize a ConvTranspose model with per-channel
weight observers until this is fixed.

Test Plan:

```
python test/test_quantization.py TestQuantizeFxOps.test_conv_transpose_1d
python test/test_quantization.py TestQuantizeFxOps.test_conv_transpose_2d
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vkuzo committed:
nit on "fx quant: hook up ConvTranspose{n}d"


Summary:

Quantization of `ConvTranpose{n}d` is supported in Eager mode. This PR
adds the support for FX graph mode.

Note: this currenlty only works in `qnnpack` because per-channel weights
are not supported by quantized conv transpose. In a future PR we should throw
an error when someone tries to quantize a ConvTranspose model with per-channel
weight observers until this is fixed.

Test Plan:

```
python test/test_quantization.py TestQuantizeFxOps.test_conv_transpose_1d
python test/test_quantization.py TestQuantizeFxOps.test_conv_transpose_2d
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D25674636](https://our.internmc.facebook.com/intern/diff/D25674636)

[ghstack-poisoned]

jerryzh168 commented on torch/quantization/fx/quantization_patterns.py:
```diff
@@ -277,6 +277,38 @@ def convert(self, quantizer: QuantizerCls, node: Node, load_arg: Callable,
 
+@register_quant_pattern(torch.nn.ConvTranspose1d)
+@register_quant_pattern(torch.nn.ConvTranspose2d)
+class ConvTransposeQuantizeHandler(QuantizeHandler):
```
can we use this: https://github.com/pytorch/pytorch/blob/master/torch/quantization/fx/quantization_patterns.py#L554
feel free to inherit as well if you want to ConvTranspose separate from the things already listed there.

vkuzo replied:
didn't know about that one, thanks

jerryzh168 commented on test/quantization/test_quantize_fx.py:
```diff
@@ -2361,6 +2362,40 @@ def test_rnn(self):
                                      [100, -155]], dtype=torch.float).unsqueeze(0).repeat(niter, 1, 1)
         self._test_rnn_impl(qconfigs, RNNDynamicModel, module_type_strs, module_types, sample_input)
 
+    def _test_conv_transpose_impl(
```
can we use checkGraphModeFxOp? https://github.com/pytorch/pytorch/blob/master/test/quantization/test_quantize_fx.py#L1368
this is the common check function we use for all modules, feel free to add more functionalities there so other tests can benefit as well.

jerryzh168 replied:
I see, we should probably add eager mode numerics check to checkGraphModeFxOp

vkuzo committed:
comments on "fx quant: hook up ConvTranspose{n}d"


Summary:

Quantization of `ConvTranpose{n}d` is supported in Eager mode. This PR
adds the support for FX graph mode.

Note: this currenlty only works in `qnnpack` because per-channel weights
are not supported by quantized conv transpose. In a future PR we should throw
an error when someone tries to quantize a ConvTranspose model with per-channel
weight observers until this is fixed.

Test Plan:

```
python test/test_quantization.py TestQuantizeFxOps.test_conv_transpose_1d
python test/test_quantization.py TestQuantizeFxOps.test_conv_transpose_2d
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D25674636](https://our.internmc.facebook.com/intern/diff/D25674636)

[ghstack-poisoned]

vkuzo committed:
Update on "fx quant: hook up ConvTranspose{n}d"


Summary:

Quantization of `ConvTranpose{n}d` is supported in Eager mode. This PR
adds the support for FX graph mode.

Note: this currenlty only works in `qnnpack` because per-channel weights
are not supported by quantized conv transpose. In a future PR we should throw
an error when someone tries to quantize a ConvTranspose model with per-channel
weight observers until this is fixed.

Test Plan:

```
python test/test_quantization.py TestQuantizeFxOps.test_conv_transpose_1d
python test/test_quantization.py TestQuantizeFxOps.test_conv_transpose_2d
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D25674636](https://our.internmc.facebook.com/intern/diff/D25674636)

[ghstack-poisoned]

vkuzo committed:
Update on "fx quant: hook up ConvTranspose{n}d"


Summary:

Quantization of `ConvTranpose{n}d` is supported in Eager mode. This PR
adds the support for FX graph mode.

Note: this currenlty only works in `qnnpack` because per-channel weights
are not supported by quantized conv transpose. In a future PR we should throw
an error when someone tries to quantize a ConvTranspose model with per-channel
weight observers until this is fixed.

Test Plan:

```
python test/test_quantization.py TestQuantizeFxOps.test_conv_transpose_1d
python test/test_quantization.py TestQuantizeFxOps.test_conv_transpose_2d
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D25674636](https://our.internmc.facebook.com/intern/diff/D25674636)

[ghstack-poisoned]

vkuzo committed:
Update on "fx quant: hook up ConvTranspose{n}d"


Summary:

Quantization of `ConvTranpose{n}d` is supported in Eager mode. This PR
adds the support for FX graph mode.

Note: this currenlty only works in `qnnpack` because per-channel weights
are not supported by quantized conv transpose. In a future PR we should throw
an error when someone tries to quantize a ConvTranspose model with per-channel
weight observers until this is fixed.

Test Plan:

```
python test/test_quantization.py TestQuantizeFxOps.test_conv_transpose_1d
python test/test_quantization.py TestQuantizeFxOps.test_conv_transpose_2d
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D25674636](https://our.internmc.facebook.com/intern/diff/D25674636)

[ghstack-poisoned]
